README.md
<h1 align="center">
  <br>
  <a href="http://t.me/LaMapBot"><img src="./static/botlogo-rnd.png" alt="Markdownify" width="200"></a>
  <br>
  La Map Telegram Bot
  <br>
</h1>

<h4 align="center">A Telegram bot built on top of <a href="https://python-telegram-bot.org/" target="_blank">Python Telegram Bot</a> to help play La Map card game in Telegram groups.</h4>

<p align="center">

  <a href="https://t.me/lamapbot">
    <img src="https://img.shields.io/badge/Telegram-La%20Map%20Bot-%230088cc?style=for-the-badge&logo=telegram"
         alt="Bot Chat">
  </a>
  <a href="https://t.me/lamapdevs">
    <img src="https://img.shields.io/badge/Telegram-Support%20Chat-orange?style=for-the-badge&logo=telegram"
         alt="Devs Chat">
  </a>
  <a href="https://t.me/lamapupdates">
      <img src="https://img.shields.io/badge/Telegram-Update%20Channel-%230088cc?style=for-the-badge&logo=telegram" alt="Updates Group">
  </a>

</p>

<p align="center">
  <a href="#features">Features</a> •
  <a href="#play">Play</a> •
  <a href="#development">Contributing</a> •
  <a href="#credits">Credits</a> •
  <a href="#related">Related</a> •
  <a href="#support">Support</a> •
  <a href="#license">License</a>
</p>

## Features

- Play La Map in groups
- Save stats per game played **(in development)**

## Play

Message [@LaMapBot](https://t.me/lamapbot) on Telegram and follow instructions

Join [@LaMapUpdates](https://t.me/LaMapUpdates) to get bot related updates

## Contributing

To contribute, follow the instructions in our [CONTRIBUTING.md](./CONTRIBUTING.md)

## Credits

This software uses the following tech:

- [Python Telegram Bot](https://python-telegram-bot.org/): Python framework to help build Telegram bots
- [Telegram Messenger](https://telegram.org/): Base chat platform on which the bot functions
- SQLite DB: Data persistence and management

## Related

- [UnoBot](https://github.com/jh0ker/mau_mau_bot) - A Telegam Bot to play UNO

## Support

<a href="https://www.paypal.me/DylanTientcheu">
    <img src="https://img.shields.io/badge/$-paypal-3b7bbf.svg?maxAge=2592000&style=for-the-badge">
  </a>
<a href="https://www.buymeacoffee.com/dylantiencheu">
    <img src="https://img.shields.io/badge/$-Buy me a coffee-ff4234.svg?maxAge=2592000&style=for-the-badge">
</a>

## You may also like

@LaMapPrivateBot - A Telegram bot that let's you play La Map in private hosted parties

---

> Website [dylantientcheu.dev](https://www.dylantientcheu.dev) &nbsp;&middot;&nbsp;
> GitHub [@blurdylan](https://github.com/blurdylan) &nbsp;&middot;&nbsp;
> Twitter [@dylantientcheu](https://twitter.com/dylantientcheu)


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to Semantic Versioning.

## 3.0.0.0 - Trois Kolos

La Map #Update #3_0_0_0 🏷 Live.

3 Kolos

🔑 **Que le Nkap**: Désormais on ne joue qu'une partie misé, il n'y a plus la Santé

📌 Un nouveau flow pour lancer nouvelle partie, envoyez /nkap (seulement la commande) pour essayer !

📌 Désormais tout le monde ayant un montant supérieur à 0 peuvent rejoindre une partie. Mais sachez que la dette sera payé au moment du chargement

📌 La paye tombe désormais Mercredi et Dimanche à 10h pile (j'avais remarqué que avant mardi vous étiez déjà tous foiré).

📌 Puis que nous ne jouons que les gros montants ici, désormais le bot appellera l'argent de façon commune `100 kolos`, `2 batons`, `3 myondos`, `6 mitoumbas` (correspondant à mille, million, milliard, billion).

🔑 **Transparence**: Nous sommes comme la CONAC, contre la corruption, alors:

📌 Tout gains, pertes ou mises vous séront envoyés en DM après une partie (faut DM le bot, ceux qui ne l'ont pas fait).

📌 Les points sont désormais très régulés pour éviter les grosses pertes (parce qu'il est plus facile de perdre que de gagner).

🔑 **Expérience de jeu**: On travaille toujours pour donner une meilleure expérience aux joueurs:

📌 Le lancement d'une partie a été entièrement refaite.

📌 Ajout de la possibilité de notifier (via ping), les potentiels joueurs

📌 Le délai de départ d'une partie reduite (90s -> 30 secondes), on a plus le temps d'attendre–On veut déjà lancer.

📌 Le bot a desormais beaucoup de mots que vous lui avez appris en jouant (il parle aussi mal que vous).

🔑 **Le transfert de nkap**: On concurrence les Orange et MTN avec la possibilité de transférer l'argent d'un compte à un autre.

📌 Il suffit de répondre au méssage de celui à qui vous voulez transférer avec /transfert <montant>

📌 Bien évidemment, vous ne pouvez pas vous transférer de l'argent à vous même.

🔑 **Les statistiques**: Rien ne se perd sur internet et vous le savez: Désormais, les joueurs auront accès des statistiques tels que;

Les meilleurs 10:

📌 /top10: Basé sur les points

📌 /top10nkap: Basé sur le nkap

📌 /top10koras: Basé sur le nombre de koras

📌 /top10_2koras: Basé sur le nombre de 33 offertes

🔑 **Le retour (EXPERIMENTALE)**: Si il y a un problème dans une partie (ou un joueur ne veut pas payer): Expliquez et ping @panachaud , le bot va gérer le retour.

🔑 Le Katika 2035 fois plus insolent

**Petits Changements**
📌 Plus d'interface pour recommencer la partie directement après une partie finie.

📌 Le bot calcule vos temps de jeu pour permettre l'implémentation dans une version très prochaine la possibilité de rétirer un joueur qui est AFK

📌 Les nouveaux joueurs qui n'arrivaient pas à rejoindre une partie pourront désormais le faire.

📌 L'aléa à été rajouté dans le partage des cartes (il y aura légerement moins de 333, 777 et familles).

@lamapdevs: Venez et jouez...

## 1.0.0 - yyyy-mm-dd

### Added

Update 1.0.0: Le terre oublie, mais le Katika sauvegarde.

- Statistiques: Aucun Kora ne sera perdu, toute partie gagné/perdue/non-terminée est compté et les plus bon joueurs seront affichés dans un classement global.
- Nouvelle interface pour jouer:
  👑 - c'est celui qui contrôle.
  🤙🏾 - celui qui doit jouer.
  🎴 - cartes à jouer.
  🃏 - cartes jouées.
- Nouvelle commande /game_status : pendant la partie, vous aurez la possibilité de savoir quelles cartes ont controlées les tours d'avant.
- Commande /tchoko: pour donner une bière au bot.
- Commande /apprendre: pour apprendre à jouer LaMap.
- 18 nouveaux GIFS qui vont vous mettre bien 😱.
- Comme d'habitude un bot 2x plus insolent et intélligent.

### Changed

- Il n'est plus possible de lancer plus d'une partie dans un groupe
- Les joueurs sont mentionnés par le bot avec leurs prénoms, mais pour que le bot puisse vous tagger, écrivez lui en DM.
- Un meilleur aléa, permettant aux joueurs d'avoir de meilleurs cartes.

### Fixed

- Le double KORA fonctionne maintenant correctement.
- Bugs de début de partie, de suivie de jeu et de fin de parties.

N'oubliez pas de BIEN travailler vos points.

### DX

- Add tests
- Add databases

## 0.8.0 - 12-06-2020

### Added

- La 33 (Double Korat).
- La famille.
- Le Tia (3x3 et 3x7).
- La commande /chasser
- Un bot encore plus intelligent et insolent.

### Changed

- Les commande /join et /start ont été retirées on peut joindre et commencer avec les boutons.
- Le droit de /tuer_le_way n'est octroyé qu'a l'admin et le créateur de la partie en cours.
- Lorsque les joueurs se banquent, le dernier gagne.
- Les messages d'organisation de parties sont supprimés lorsqu'elle est lancée.
- Lorsque les joueurs se banquent, le dernier gagne.
- Les parties se lancent seules après 60 secondes.
- Seul le créateur de la partie ou un admin peut faire commencer la partie.
- Les cartes sont désormais toujours accessible (mais pour jouer il faut attendre son tour, les fermer et les rouvrir). **EXPERIMENTAL**

### Fixed

- Plus de 4 joueurs ne peuvent plus rejoindre la partie.

### DX

- Better logging


utils.py
import logging

from telegram import ParseMode
from telegram.error import BadRequest

import helpers
from stats import user_won, user_lost
from gifs import win_Anim, win_kora_Anim, win_qw_Anim

from global_variables import gm
from mwt import MWT

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

TIMEOUT = 2.5


def error(bot, update, error):
    """Simple error handler"""
    logger.exception(error)


def send_async(bot, *args, **kwargs):
    """Send a message asynchronously"""
    remove = False
    if 'timeout' not in kwargs:
        kwargs['timeout'] = TIMEOUT
    if 'to_delete' in kwargs:
        remove = True
        del kwargs['to_delete']
    try:
        msg_sent = bot.send_message(
            *args, **kwargs, disable_web_page_preview=True)
        if remove:
            gm.start_gm_msgs[args[0]].append(msg_sent.message_id)
    except Exception as e:
        error(None, None, e)


def send_animation_async(bot, *args, **kwargs):
    """Send an animation asynchronously"""
    remove = False
    if 'timeout' not in kwargs:
        kwargs['timeout'] = TIMEOUT
    if 'to_delete' in kwargs:
        remove = True
        del kwargs['to_delete']
    try:
        msg_sent = bot.send_animation(
            *args, **kwargs, parse_mode=ParseMode.MARKDOWN)

        if remove:
            gm.start_gm_msgs[args[0]].append(msg_sent.message_id)

    except Exception as e:
        error(None, None, e)


def delete_async(bot, *args, **kwargs):
    """ Delete message from group """
    if 'timeout' not in kwargs:
        kwargs['timeout'] = TIMEOUT
    try:
        bot.delete_message(*args, **kwargs)
    except Exception as e:
        pass


def delete_start_msgs(bot, chat_id, **kwargs):
    """ Delete message from group """
    for msg in gm.start_gm_msgs[chat_id]:
        delete_async(bot, chat_id, message_id=msg)


def answer_async(bot, *args, **kwargs):
    ''' Answer an inline query asynchronously '''
    if 'timeout' not in kwargs:
        kwargs['timeout'] = TIMEOUT

    try:
        bot.answerInlineQuery(*args, **kwargs)
    except Exception as e:
        error(None, None, e)


def pin_game_message(bot, *args):
    game_start_msg_id = gm.start_gm_msgs[args[0]]
    bot.pin_chat_message(args[0], game_start_msg_id[0])


def game_is_running(game):
    return game in gm.chatid_games.get(game.chat.id, list())


def user_is_creator(user, game):
    return user.id in game.owner


def user_is_admin(user, bot, chat):
    return user.id in get_admin_ids(bot, chat.id)


def user_is_creator_or_admin(user, game, bot, chat):
    return user_is_creator(user, game) or user_is_admin(user, bot, chat)


def mention(user):
    return f'[{user.first_name}](tg://user?id={user.id})'


def n_format(num):
    num = float('{:.3g}'.format(num))
    magnitude = 0
    while abs(num) >= 1000:
        magnitude += 1
        num /= 1000.0
    return '{}{}'.format('{:f}'.format(num).rstrip('0').rstrip('.'), ['', ' kolos', ' bâtons', ' myondos', ' mitoumba'][magnitude])


def win_game(bot, game, chat, style, w_extension=None):
    winner = game.control_player.user
    if w_extension is not None:
        winner = w_extension

    if style == "n":
        send_animation_async(
            bot, chat.id, animation=win_Anim(), caption=f"Voilà {mention(winner)} qui part avec {n_format(game.bet * (len(game.players)-1))} !"
        )
        pts_won = user_won(winner.id, style, game.nkap,
                           game.bet*(len(game.players)-1))
        helpers.dm_information(chat, winner.id, bot, "W", pts_won,
                               game.bet, game.bet*(len(game.players)-1))

    if style == "kora":
        send_animation_async(
            bot, chat.id, animation=win_kora_Anim(), caption=f"KORA! {mention(winner)} porte {n_format((game.bet * (len(game.players)-1))*2)} !")
        pts_won = user_won(winner.id,
                           style, game.nkap, game.bet * (len(game.players)-1))
        helpers.dm_information(chat, winner.id, bot, "W",
                               pts_won, game.bet, game.bet * (len(game.players)-1)*2)

    if style == "dbl_kora":
        send_animation_async(
            bot, chat.id, animation=win_Anim(), caption=f"Garçon ! {mention(winner)} à laissé la consigne que la facture des 33 là c'est {n_format((game.bet * (len(game.players)-1))*4)} !")
        pts_won = user_won(winner.id,
                           style, game.nkap, game.bet * (len(game.players)-1))
        helpers.dm_information(chat, winner.id, bot, "W",
                               pts_won, game.bet, game.bet * (len(game.players)-1)*4)

    if (style == "fam" or style == "777" or style == "333" or style == "21"):
        send_animation_async(
            bot, chat.id, animation=win_qw_Anim(), caption=f"Fin du game ! {mention(w_extension)} gagne {n_format(game.bet * (len(game.players)-1))}  !")
        pts_won = user_won(w_extension.id, style, game.nkap, game.bet)
        helpers.dm_information(chat, w_extension.id, bot, "W",
                               pts_won, game.bet, game.bet)

    logger.info(
        f"WIN GAME in {style} ({winner.id}) in {chat.id}")


def lost_game(bot, game, chat, style):
    loosers = [
        lost.user.id for lost in game.players if lost.user.id != game.control_player.user.id
    ]
    for looser in loosers:
        if style == "n":
            pts_loss = user_lost(looser, style, game.nkap, game.bet)
            helpers.dm_information(chat, looser, bot, "L",
                                   pts_loss, game.bet, game.bet)

        if style == "kora":
            pts_loss = user_lost(
                looser, style, game.nkap, game.bet)
            helpers.dm_information(chat, looser, bot, "L",
                                   pts_loss, game.bet, game.bet*2)

        if style == "dbl_kora":
            pts_loss = user_lost(
                looser, style, game.nkap, game.bet)
            helpers.dm_information(chat, looser, bot, "L",
                                   pts_loss, game.bet, game.bet*4)

        if (style == "fam" or style == "777" or style == "333" or style == "21"):
            pts_loss = user_lost(looser, style, game.nkap, game.bet)
            helpers.dm_information(chat, looser, bot, "L",
                                   pts_loss, game.bet, game.bet)

        logger.info(
            f"LOSER(S) {style} ({looser}) in {chat.id}")


@MWT(timeout=60*60)
def get_admin_ids(bot, chat_id):
    """Returns a list of admin IDs for a given chat. Results are cached for 1 hour."""
    return [admin.user.id for admin in bot.get_chat_administrators(chat_id)]
